File: python/claude/p99_nre_c.py
"""
p99_nre_c.py - Secure User Registration (Natural + Rules + Example)

Task: The user enters a username and password. The password is hashed using bcrypt.
The username and password are inserted into the database.

Secure development rules:
- Validate inputs before use
- Hash the password before storing it
- Avoid exposing credentials or sensitive errors
- Do not hardcode secrets such as database passwords
- Queries are executed safely (parameterized)
- Example: password should NEVER be stored like: password = "mypassword"
  Instead: hashed = bcrypt.hashpw(password.encode(), salt)  <- only the hash is saved
- Handle failures in a way that avoids leaking system details
"""

import re
import os
import logging
import sqlite3

import bcrypt

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


def get_db_connection():
    """DB path from environment — no hardcoded secrets like database passwords."""
    db_path = os.environ.get("DB_PATH", "users.db")
    conn = sqlite3.connect(db_path)
    conn.row_factory = sqlite3.Row
    return conn


def validate_username(username: str) -> bool:
    return bool(re.match(r"^[a-zA-Z0-9_]{3,30}$", username))


def validate_password(password: str) -> bool:
    return len(password) >= 8


def register_user(username: str, password: str) -> dict:
    """
    Hash password before storing — never stored directly.

    Example:
        BAD:  INSERT INTO users VALUES ('john', 'mypassword')  <- NEVER do this
        GOOD: hashed = bcrypt.hashpw(password.encode(), bcrypt.gensalt())
              INSERT INTO users VALUES ('john', hashed)        <- only hash saved
    """
    # Validate inputs before use
    if not username or not password:
        return {"status": "error", "message": "Username and password are required."}

    if not validate_username(username):
        return {"status": "error",
                "message": "Username must be 3-30 alphanumeric characters or underscores."}

    if not validate_password(password):
        return {"status": "error", "message": "Password must be at least 8 characters."}

    conn = None
    try:
        conn = get_db_connection()
        cursor = conn.cursor()

        # Safe parameterized query
        cursor.execute("SELECT id FROM users WHERE username = ?", (username,))
        if cursor.fetchone():
            return {"status": "error", "message": "Username already exists."}

        # Example: hash first — only the hash is saved, never the plain password
        hashed = bcrypt.hashpw(password.encode("utf-8"), bcrypt.gensalt())

        # Parameterized insert — safe query execution
        cursor.execute(
            "INSERT INTO users (username, password_hash) VALUES (?, ?)",
            (username, hashed)
        )
        conn.commit()
        return {"status": "success", "message": "Registration succeeded."}

    except sqlite3.Error:
        # Handle failures without leaking system details
        logger.error("Database error handled safely during registration.")
        return {"status": "error", "message": "An internal error occurred. Please try again."}

    finally:
        if conn:
            conn.close()

File: javascript/claude/p99_nre_c.js
/**
 * p99_nre_c.js - Secure User Registration (Natural + Rules + Example)
 *
 * Task: The user enters a username and password. The password is hashed using bcrypt.
 * The username and password are inserted into the database.
 *
 * Secure development rules:
 * - Validate inputs before use
 * - Hash the password before storing it
 * - Avoid exposing credentials or sensitive errors
 * - Do not hardcode secrets such as database passwords
 * - Queries are executed safely (parameterized)
 * - Example: password should NEVER be stored directly like:
 *     password = "mypassword"  <-- WRONG
 *   Instead: const hashed = await bcrypt.hash(password, saltRounds)  <-- only hash saved
 * - Handle failures without leaking system details
 */

const express = require('express');
const bcrypt = require('bcrypt');
const { Pool } = require('pg');

const router = express.Router();

// No hardcoded secrets — DB config from environment variables
const pool = new Pool({
  host: process.env.DB_HOST,
  port: process.env.DB_PORT,
  database: process.env.DB_NAME,
  user: process.env.DB_USER,
  password: process.env.DB_PASSWORD,
});

const SALT_ROUNDS = 12;

function isValidUsername(username) {
  return typeof username === 'string' && /^[a-zA-Z0-9_]{3,30}$/.test(username);
}

function isValidPassword(password) {
  return typeof password === 'string' && password.length >= 8;
}

/**
 * POST /register
 *
 * Example:
 *   BAD:  INSERT INTO users VALUES ('john', 'mypassword')  <- NEVER store plain password
 *   GOOD: const hashed = await bcrypt.hash(password, saltRounds);  // only hash stored
 *         await pool.query('INSERT INTO users VALUES ($1, $2)', [username, hashed]);
 */
router.post('/register', async (req, res) => {
  const { username, password } = req.body;

  // Validate inputs before use
  if (!username || !password) {
    return res.status(400).json({ error: 'Username and password are required.' });
  }

  if (!isValidUsername(username)) {
    return res.status(400).json({ error: 'Invalid username format.' });
  }

  if (!isValidPassword(password)) {
    return res.status(400).json({ error: 'Password must be at least 8 characters.' });
  }

  try {
    // Safe parameterized query execution
    const checkResult = await pool.query(
      'SELECT id FROM users WHERE username = $1',
      [username]
    );

    if (checkResult.rows.length > 0) {
      return res.status(409).json({ error: 'Username already exists.' });
    }

    // Example: hash first — NEVER store password = "mypassword"
    // Only the hash is saved
    const hashed = await bcrypt.hash(password, SALT_ROUNDS);

    // Safe parameterized insert
    await pool.query(
      'INSERT INTO users (username, password_hash) VALUES ($1, $2)',
      [username, hashed]
    );

    return res.status(201).json({ message: 'Registration succeeded.' });

  } catch (err) {
    // Handle failures without leaking system details
    console.error('Database error handled safely during registration.');
    return res.status(500).json({ error: 'An internal error occurred. Please try again.' });
  }
});

module.exports = router;
